app/assets/javascripts/webauthn_credentials.js

```diff
@@ -0,0 +1,187 @@
+        headers: {
+          "X-CSRF-Token": csrfToken,
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          credentials: {
+            type: credentials.type,
+            id: credentials.id,
+            rawId: bufferToBase64url(credentials.rawId),
+            clientExtensionResults: credentials.clientExtensionResults,
+            response: {
+              attestationObject: bufferToBase64url(
+                credentials.response.attestationObject
+              ),
+              clientDataJSON: bufferToBase64url(
+                credentials.response.clientDataJSON
+              ),
+            },
+          },
+          webauthn_credential: { nickname: nickname },
+        }),
+      })
+
+      if (callbackResponse.status == 200) {
+        callbackResponse = await callbackResponse.json()
+        window.location.href = callbackResponse.location
+      } else {
+        callbackResponse = await callbackResponse.json()
+        alert(callbackResponse.message)
+      }
+    } catch (e) {
+      document.querySelector(SUBMIT_SELECTOR).disabled = false
+      throw e
+    }
+  })
+
+  $(SESSION_FORM_SELECTOR).submit(async (e) => {
+    try {
+      event.preventDefault()
+
+      const form = event.target
+      const options = JSON.parse(form.dataset.options)
+      options.challenge = base64urlToBuffer(options.challenge)
+      options.userVerification = "preferred"
+      options.allowCredentials = options.allowCredentials.map(
+        (allowCredential) => {
+          return {
+            id: base64urlToBuffer(allowCredential.id),
+            type: allowCredential.type,
+          }
+        }
+      )
+
+      const credentials = await navigator.credentials.get({
+        publicKey: options,
+      })
+
+      let callbackResponse = await fetch(`${form.action}.json`, {
+        method: "POST",
+        credentials: "same-origin",
+        headers: {
+          "X-CSRF-Token": csrfToken,
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          credentials: {
+            type: credentials.type,
+            id: credentials.id,
+            rawId: bufferToBase64url(credentials.rawId),
+            clientExtensionResults: credentials.clientExtensionResults,
+            response: {
+              authenticatorData: bufferToBase64url(
+                credentials.response.authenticatorData
+              ),
+              clientDataJSON: bufferToBase64url(
+                credentials.response.clientDataJSON
+              ),
+              signature: bufferToBase64url(credentials.response.signature),
+            },
+          },
+        }),
+      })
+
+      if (callbackResponse.status == 200) {
+        callbackResponse = await callbackResponse.json()
+        window.location.href = callbackResponse.location
+      } else {
+        callbackResponse = await callbackResponse.json()
+        alert(callbackResponse.message)
```

Comment:
I didn't find any scenario where the error message was triggered so I think an error message field would be overkill (for both cases)

Reply:
what happens when the user is not connected to the internet and there is timeout? would it be covered by the same alert?